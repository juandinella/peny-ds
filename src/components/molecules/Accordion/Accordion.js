import React, { useState, useRef } from 'react'
import PropTypes from 'prop-types'
import styles from './Accordion.styles'
import Skeleton from '../../atoms/Skeleton'
import Utils from '../../../utils'
import Text from '../../atoms/Text'
import Spacer from '../../layout/Spacer'
import Container from '../../layout/Container'

/**
 * Accordion
 */

const Accordion = ({ id, content, title, shouldRender }) => {
  const [active, setActive] = useState('')
  const [height, setHeight] = useState('0px')

  const contentRef = useRef(null)

  const toggleAccordion = () => {
    setActive(active === '' ? 'is-open' : '')
    setHeight(
      active === 'is-open' ? '0px' : `${contentRef.current.scrollHeight}px`
    )
  }

  return (
    <Skeleton
      id={`${id}_skeleton`}
      show={!Utils.emptyValue(shouldRender) && !shouldRender()}
    >
      <div css={styles} id={id} data-testid={id}>
        <button
          className={`accordion-toggle ${active}`}
          onClick={toggleAccordion}
          aria-expanded={active ? 'true' : 'false'}
          id={`${id}_button`}
          data-testid={`${id}_button`}
        >
          <Container flex alignItems="center" justifyContent="flex-start">
            <Container flex direction="column" alignItems="flex-start">
              <Text size="md" id={`${id}_title`}>
                {title}
              </Text>
            </Container>
          </Container>
          <div className="accordion-chevron">
            <svg
              width="16"
              height="16"
              viewBox="0 0 16 16"
              role="img"
              fill="none"
              xmlns="http://www.w3.org/2000/svg"
              aria-labelledby="IconChevronDown-label"
              aria-hidden="true"
              focusable="false"
            >
              <title id="IconChevronDown-label">ChevronDown</title>
              <path d="M13.65 4.643a.5.5 0 01.7.714L8 11.594 1.65 5.357a.5.5 0 01.7-.714L8 10.193l5.65-5.55z"></path>
            </svg>
          </div>
        </button>
        <section
          ref={contentRef}
          style={{ maxHeight: `${height}` }}
          className="accordion-content"
          aria-hidden={active ? 'false' : 'true'}
        >
          <div id={`${id}_content`} data-testid={`${id}_content`}>
            {content}
          </div>
        </section>
      </div>
    </Skeleton>
  )
}
Accordion.displayName = 'Accordion'

Accordion.propTypes = {
  /** Accordion's id */
  id: PropTypes.string,
  /** Accordion's title */
  title: PropTypes.string,
  /** Accordion's content */
  content: PropTypes.node,
  /**
   * Function that decides if the component should be rendered.
   *
   * @return {boolean} false if component should not be rendered and show skeleton instead. Component renders otherwise.
   */
  shouldRender: PropTypes.func,
}

Accordion.defaultProps = {}

export default Accordion
