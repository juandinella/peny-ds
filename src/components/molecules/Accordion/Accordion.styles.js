import { css } from 'styled-components'

import layout from '../../../styles/theme/layout'

export default css`
  display: flex;
  flex-direction: column;
  width: 100%;
  background-color: var(--color-neutral-00);
  position: relative;
  box-shadow: var(--shadow-sm);
}

  &:first-of-type {
    border-radius: var(--radius-sm) var(--radius-sm) 0 0;
  }

  &:last-of-type {
    border-radius: 0 0 var(--radius-sm) var(--radius-sm);
  }

  &:after {
    content: '';
    width: calc(100% - 64px);
    margin: 0 auto;
    height: 1px;
    position: absolute;
    bottom: 0;
    left: 0;
    right: 0;
    background-color: var(--color-neutral-400);
  }

  &:last-of-type {
    &:after {
      content: none;
    }
  }

  .accordion-toggle {
    padding: var(--spacing-md) var(--spacing-sm);
    display: flex;
    align-items: center;
    justify-content: space-between;
    background: transparent;

    @media screen and (min-width: ${() => layout.breakpoints.sm}) {
      padding: var(--spacing-md) var(--spacing-lg);
    }

    .accordion-chevron {
      transition: transform var(--transition-timing-easing);

      svg path {
        fill: var(--color-primary-main);
      }
    }

    &.is-open {
      .accordion-chevron {
        transform: rotate(180deg);
      }
    }
  }

  .accordion-content {
    overflow: auto;
    transition: max-height var(--transition-timing-easing);

    > div {
      padding: 0 var(--spacing-lg) var(--spacing-md) var(--spacing-lg);
    }
  }
`
