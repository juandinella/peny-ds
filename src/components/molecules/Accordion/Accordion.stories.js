import Accordion from './Accordion'
import { getTemplate } from '../../../helpers/storybook'

const Template = getTemplate(Accordion)

export default {
  title: 'Molecules/Accordion',
  component: Accordion,
  id: 'test',
  args: {
    title: 'Accordion Title',
    content: 'Accordion content',
    id: 'accordion_id',
  },
  argTypes: {},
}

export const Default = Template.bind({})

export const Skeleton = Template.bind({})
Skeleton.args = { shouldRender: () => false }
