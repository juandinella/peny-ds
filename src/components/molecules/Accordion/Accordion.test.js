import React from 'react'

import { cleanup, render, fireEvent } from '@testing-library/react'

import Accordion from './Accordion'

import initStoryshots from '@storybook/addon-storyshots'

initStoryshots({
  storyKindRegex: /^Molecules\/Accordion$/,
})

const textContent = 'Accordion content'

const componentProps = {
  title: 'Accordion Title',
  content: <span>{textContent}</span>,
  id: 'accordion_id',
}

afterEach(cleanup)

describe('<Accordion />', () => {
  test('should render accordion correctly', () => {
    const { queryByTestId } = render(<Accordion {...componentProps} />)

    expect(queryByTestId(componentProps.id)).toBeTruthy()
  })

  test('should render skeleton', () => {
    const { queryByTestId } = render(
      <Accordion {...componentProps} shouldRender={() => false} />
    )

    expect(queryByTestId(`${componentProps.id}_skeleton`)).toBeTruthy()
  })

  test('should show title correctly', () => {
    const { queryByTestId } = render(<Accordion {...componentProps} />)

    expect(queryByTestId(`${componentProps.id}_title`)).toBeTruthy()
  })

  test('should show content correctly', () => {
    const { queryByTestId } = render(<Accordion {...componentProps} />)

    const content = queryByTestId(`${componentProps.id}_content`)

    expect(content.firstChild).toBeTruthy()
    expect(content.firstChild.textContent).toEqual(textContent)
  })

  test('should open content correctly', () => {
    const { container, queryByTestId } = render(
      <Accordion {...componentProps} />
    )

    const button = queryByTestId(`${componentProps.id}_button`)

    expect(button.className).not.toContain('is-open')
    expect(
      container.querySelector('.accordion-content').getAttribute('aria-hidden')
    ).toEqual('true')

    fireEvent.click(button)

    expect(button.className).toContain('is-open')
    expect(
      container.querySelector('.accordion-content').getAttribute('aria-hidden')
    ).toEqual('false')
  })
})
