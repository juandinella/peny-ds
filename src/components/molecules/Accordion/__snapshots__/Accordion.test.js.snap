// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Molecules/Accordion Default 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <div
    className="sc-fqcnox-0 llbqsw"
    data-testid="accordion_id"
    id="accordion_id"
  >
    <button
      aria-expanded="false"
      className="accordion-toggle "
      data-testid="accordion_id_button"
      id="accordion_id_button"
      onClick={[Function]}
    >
      <div
        className="sc-kazc2j-0 dtNeUd"
        style={
          Object {
            "bottom": undefined,
            "left": undefined,
            "right": undefined,
            "top": undefined,
          }
        }
      >
        <div
          className="sc-kazc2j-0 hHJZsJ"
          direction="column"
          style={
            Object {
              "bottom": undefined,
              "left": undefined,
              "right": undefined,
              "top": undefined,
            }
          }
        >
          <p
            className="sc-15mqapg-0 eeHXJt"
            color="neutral-900"
            data-testid="accordion_id_title"
            id="accordion_id_title"
            size="md"
          >
            Accordion Title
          </p>
        </div>
      </div>
      <div
        className="accordion-chevron"
      >
        <svg
          aria-hidden="true"
          aria-labelledby="IconChevronDown-label"
          fill="none"
          focusable="false"
          height="16"
          role="img"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="IconChevronDown-label"
          >
            ChevronDown
          </title>
          <path
            d="M13.65 4.643a.5.5 0 01.7.714L8 11.594 1.65 5.357a.5.5 0 01.7-.714L8 10.193l5.65-5.55z"
          />
        </svg>
      </div>
    </button>
    <section
      aria-hidden="true"
      className="accordion-content"
      style={
        Object {
          "maxHeight": "0px",
        }
      }
    >
      <div
        data-testid="accordion_id_content"
        id="accordion_id_content"
      >
        Accordion content
      </div>
    </section>
  </div>
</div>
`;

exports[`Storyshots Molecules/Accordion Skeleton 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <div
    className="sc-kjnynt-0 jfZhXD"
    data-testid="accordion_id_skeleton"
    id="accordion_id_skeleton"
    style={
      Object {
        "height": undefined,
        "width": undefined,
      }
    }
  >
    <div
      className="sc-fqcnox-0 llbqsw"
      data-testid="accordion_id"
      id="accordion_id"
    >
      <button
        aria-expanded="false"
        className="accordion-toggle "
        data-testid="accordion_id_button"
        id="accordion_id_button"
        onClick={[Function]}
      >
        <div
          className="sc-kazc2j-0 dtNeUd"
          style={
            Object {
              "bottom": undefined,
              "left": undefined,
              "right": undefined,
              "top": undefined,
            }
          }
        >
          <div
            className="sc-kazc2j-0 hHJZsJ"
            direction="column"
            style={
              Object {
                "bottom": undefined,
                "left": undefined,
                "right": undefined,
                "top": undefined,
              }
            }
          >
            <p
              className="sc-15mqapg-0 eeHXJt"
              color="neutral-900"
              data-testid="accordion_id_title"
              id="accordion_id_title"
              size="md"
            >
              Accordion Title
            </p>
          </div>
        </div>
        <div
          className="accordion-chevron"
        >
          <svg
            aria-hidden="true"
            aria-labelledby="IconChevronDown-label"
            fill="none"
            focusable="false"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title
              id="IconChevronDown-label"
            >
              ChevronDown
            </title>
            <path
              d="M13.65 4.643a.5.5 0 01.7.714L8 11.594 1.65 5.357a.5.5 0 01.7-.714L8 10.193l5.65-5.55z"
            />
          </svg>
        </div>
      </button>
      <section
        aria-hidden="true"
        className="accordion-content"
        style={
          Object {
            "maxHeight": "0px",
          }
        }
      >
        <div
          data-testid="accordion_id_content"
          id="accordion_id_content"
        >
          Accordion content
        </div>
      </section>
    </div>
  </div>
</div>
`;

exports[`Storyshots Molecules/Accordion With Subtitle 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <div
    className="sc-fqcnox-0 llbqsw"
    data-testid="accordion_id"
    id="accordion_id"
  >
    <button
      aria-expanded="false"
      className="accordion-toggle "
      data-testid="accordion_id_button"
      id="accordion_id_button"
      onClick={[Function]}
    >
      <div
        className="sc-kazc2j-0 dtNeUd"
        style={
          Object {
            "bottom": undefined,
            "left": undefined,
            "right": undefined,
            "top": undefined,
          }
        }
      >
        <div
          className="sc-kazc2j-0 hHJZsJ"
          direction="column"
          style={
            Object {
              "bottom": undefined,
              "left": undefined,
              "right": undefined,
              "top": undefined,
            }
          }
        >
          <p
            className="sc-15mqapg-0 eeHXJt"
            color="neutral-900"
            data-testid="accordion_id_title"
            id="accordion_id_title"
            size="md"
          >
            Accordion Title
          </p>
        </div>
      </div>
      <div
        className="accordion-chevron"
      >
        <svg
          aria-hidden="true"
          aria-labelledby="IconChevronDown-label"
          fill="none"
          focusable="false"
          height="16"
          role="img"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="IconChevronDown-label"
          >
            ChevronDown
          </title>
          <path
            d="M13.65 4.643a.5.5 0 01.7.714L8 11.594 1.65 5.357a.5.5 0 01.7-.714L8 10.193l5.65-5.55z"
          />
        </svg>
      </div>
    </button>
    <section
      aria-hidden="true"
      className="accordion-content"
      style={
        Object {
          "maxHeight": "0px",
        }
      }
    >
      <div
        data-testid="accordion_id_content"
        id="accordion_id_content"
      >
        Accordion content
      </div>
    </section>
  </div>
</div>
`;
